### ISSUE TITLE
Recommendation for concurrent operations on the same file

### ISSUE THREAD

**Issue by brandongregoryscott** [dsherret/ts-simple-ast#1270]
Hey there! I'm running into a concurrency issues with some transformations I'm doing, and I'm wondering if there's a better way to lock or refresh a file when async transformations are happening. 

Essentially, my tool works like this:

```
type RuleFunction = (file: SourceFile) => Promise<RuleResult>;
const rules: RuleFunction[] = [...];
const files: SourceFile[] = project.getSourceFiles();

const results: RuleResult[][] = await Promise.all(
  rules.map((rule) => files.map((file) => rule(file)))
);

console.log(flatten(results));
```

This has gotten me pretty far (and works pretty quickly), but I'm now running into an issue where rules are working on similar nodes (export declarations). Rule A might be changing a default export to a named export, and Rule B might be doing the same - and attempting to move the export to the end of a file. Ideally, one or the other would able to run on the file, do its work, and then let the next rule pick up where it left off - however, right now, it seems like they're both performing their operations on the original version of the file and leaving behind two exports which is resulting in a duplicate identifier error.

Long story short, I'm wondering what the best strategy for maintaining speed & safe concurrency for a system like this would be (and if there's any built-in APIs aside from those that I know of like `Project.save()`, `SourceFile.save()` or `SourceFile.refreshFromFileSystem()` which might help out!)

**Comment by brandongregoryscott** [comment]
Update: I implemented a very crude locking system per file/rule and narrowed it down a bit further - I think it's actually an issue with my own transformation code not checking to see if the export might exist before adding a new one 🤦 Going to close this out.
